Use AWS_CRT_MEMORY_TRACING to enable memory trace

marcoemorais-aws opened the pull request:
* Prior to this commit, the AWS CRT resource manager
  (SharedCrtResourceManager) in the device client enabled
  memory trace for all allocations without any means to
  disable this behavior.
* On this commit, memory trace is made configurable using
  the AWS_CRT_MEMORY_TRACING environment variable.
* By default, memory trace is disabled.
* Setting the value of AWS_CRT_MEMORY_TRACING=1 (BYTES)
  enables tracking of the size of the allocated memory.
* Setting the value of AWS_CRT_MEMORY_TRACING=2 (STACKS)
  enables capturing the callstack of each allocation.
* Device client unit tests in the CI pipeline are invoked
  with AWS_CRT_MEMORY_TRACING=1.

### Motivation
- Please give a brief description for the background of this change.
- Issue number: 


### Modifications
#### Change summary
Please describe what changes are included in this pull request. 

#### Revision diff summary
If there is more than one revision, please explain what has been changed since the last revision.

### Testing
 **Is your change tested? If not, please justify the reason.**  
 **Please list your testing steps and test results.** 
- CI test run result: <link>


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

HarshGandhi-AWS commented on .github/build.sh:
```diff
@@ -231,4 +231,4 @@ fi
 
 ### Run Tests ###
-./test/test-aws-iot-device-client
+env AWS_CRT_MEMORY_TRACING=1 ./test/test-aws-iot-device-client
```
Why are we not adding tests for `AWS_CRT_MEMORY_TRACING=2 OR 0`?

marcoemorais-aws replied:
You are referring to tests that the configuration value is interpreted from environment correctly rather than tests of the tracing capability right? I agree with the former and will add.

marcoemorais-aws commented on test/CMakeLists.txt:
```diff
@@ -109,3 +109,4 @@ if (LINK_DL)
 endif ()
 
 add_test(${GTEST_PROJECT} ${GTEST_PROJECT})
+set_tests_properties(${GTEST_PROJECT} PROPERTIES ENVIRONMENT AWS_CRT_MEMORY_TRACING=1)
```
ENV variable is always a runtime setting. In other words, setting the property at build time does not mean that you are "baking" the value into the executable so I don't think there is any potential for misunderstanding.

I didn't add anything to `setup.sh` for the same reasons. The behavior of `AWS_CRT_MEMORY_TRACING` is identical to `AWSIOT_TUNNEL_ACCESS_TOKEN` which also doesn't appear in `setup.sh`.

I agree the variable should be added to README and will add mention of both `AWS_CRT_MEMORY_TRACING` and `AWSIOT_TUNNEL_ACCESS_TOKEN`.

HarshGandhi-AWS replied:
Sounds good. And since you are setting `AWS_CRT_MEMORY_TRACING` value for only unit tests so it makes sense to not to add it in `setup.sh` script.

graebm reviewed (commented):
agree with Harsh, but otherwise looks good

marcoemorais-aws commented:
@HarshGandhi-AWS @graebm Updated the review with the following additional changes:
* Added a new documentation page `docs/ENV.md` with a description of all environment variables including memory trace.
* Added a unit test for reading the value of the trace environment variables into `PlainConfig`.
* Added some logic to the device client main to support printing the contents of the trace when `SIGHUP` is received.  If there are no allocations which are not freed, then nothing is printed.

HarshGandhi-AWS reviewed (approved):
Looks good to me.
